token-client/src/main/java/com/sap/cloud/security/xsuaa/client/DefaultOAuth2TokenService.java

```diff
@@ -0,0 +1,201 @@
+package com.sap.cloud.security.xsuaa.client;
+
+import org.springframework.http.*;
+import org.springframework.lang.NonNull;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestOperations;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.sap.cloud.security.xsuaa.client.OAuth2TokenServiceConstants.*;
+import static com.sap.cloud.security.xsuaa.client.OAuth2TokenServiceConstants.GRANT_TYPE;
+import static com.sap.cloud.security.xsuaa.client.OAuth2TokenServiceConstants.GRANT_TYPE_CLIENT_CREDENTIALS;
+
+public class DefaultOAuth2TokenService implements OAuth2TokenService {
+
+	private RestOperations restOperations;
+
+	public DefaultOAuth2TokenService(@NonNull RestOperations restOperations) {
+		Assert.notNull(restOperations, "restOperations is required");
+		this.restOperations = restOperations;
+	}
+
+	@Override
+	public OAuth2AccessToken retrieveAccessTokenViaClientCredentialsGrant(@NonNull URI tokenEndpointUri,
+			@NonNull ClientCredentials clientCredentials,
+			@Nullable Map<String, String> optionalParameters) throws OAuth2ServiceException {
+		Assert.notNull(tokenEndpointUri, "tokenEndpointUri is required");
+		Assert.notNull(clientCredentials, "clientCredentials is required");
+
+		Map<String, String> parameters = copy(optionalParameters);
+		parameters.put(GRANT_TYPE, GRANT_TYPE_CLIENT_CREDENTIALS);
+
+		// build header
+		HttpHeaders headers = createHeadersWithAuthorization(clientCredentials);
+
+		return requestAccessToken(tokenEndpointUri, headers, parameters);
+	}
+
+	@Override
+	public OAuth2AccessToken retrieveAccessTokenViaUserTokenGrant(@NonNull URI tokenEndpointUri,
+			@NonNull ClientCredentials clientCredentials, @NonNull String token,
+			@Nullable Map<String, String> optionalParameters)
+			throws OAuth2ServiceException {
+		Assert.notNull(tokenEndpointUri, "tokenEndpointUri is required");
+		Assert.notNull(clientCredentials, "clientCredentials is required");
+		Assert.notNull(token, "token is required");
+
+		Map<String, String> parameters = copy(optionalParameters);
+		parameters.put(GRANT_TYPE, GRANT_TYPE_USER_TOKEN);
+		parameters.put(PARAMETER_CLIENT_ID, clientCredentials.getId());
+
+		// build header
+		HttpHeaders headers = createHeadersWithAuthorization(token);
```

Comment:
authz header see above

Reply:
thanks a lot for this hint. Its done with this commit:
https://github.com/SAP/cloud-security-xsuaa-integration/pull/124/commits/44bc95c7e5452af60373e0984f278f3d38d674ea